Add an otelcol.processor.attributes component (#3349)
grafana/agent: docs/sources/flow/reference/components/otelcol.processor.attributes.md

components and modifies attributes of a span, log, or metric.
It also supports the ability to filter and match input data to determine if 
it should be included or excluded for attribute modifications.

> **NOTE**: `otelcol.processor.attributes` is a wrapper over the upstream
> OpenTelemetry Collector `attributes` processor. Bug reports or feature requests
> will be redirected to the upstream repository, if necessary.

Multiple `otelcol.processor.attributes` components can be specified by giving them
different labels.

## Usage

```river
otelcol.processor.attributes "LABEL" {
  output {
    metrics = [...]
    logs    = [...]
    traces  = [...]
  }
}
```

## Arguments

`otelcol.processor.attributes` doesn't support any arguments and is configured fully
through inner blocks.

## Blocks

The following blocks are supported inside the definition of
`otelcol.processor.attributes`:
Hierarchy | Block | Description | Required
--------- | ----- | ----------- | --------
output | [output][] | Configures where to send received telemetry data. | yes
action | [action][] | Actions to take on the attributes of incoming metrics/logs/traces. | no

> include | [include/exclude] | Filter for data to be included to this processor's actions. | no
> include > regexp | [regexp] | Regex cache settings. | no
> include > attribute | [attribute] | A list of attributes to match against. | no
> include > resource | [resource] | A list of items to match the resources against. | no
> include > library | [library] | A list of items to match the implementation library against. | no
> include > log_severity_number | [library] | How to match against a log record's SeverityNumber, if defined. | no
> exclude | [include/exclude] | Filter for data to be excluded from this processor's actions | no
> exclude > regexp | [regexp] | Regex cache settings. | no
> exclude > attribute | [attribute] | A list of attributes to match against. | no
> exclude > resource | [resource] | A list of items to match the resources against. | no
> exclude > library | [library] | A list of items to match the implementation library against. | no
> exclude > log_severity_number | [library] | How to match against a log record's SeverityNumber, if defined. | no

Reviewer: ```suggestion
include | [include/exclude][] | Filter for data to be included to this processor's actions. | no
include > regexp | [regexp][] | Regex cache settings. | no
include > attribute | [attribute][] | A list of attributes to match against. | no
include > resource | [resource][] | A list of items to match the resources against. | no
include > library | [library][] | A list of items to match the implementation library against. | no
include > log_severity_number | [library][] | How to match against a log record's SeverityNumber, if defined. | no
exclude | [include/exclude][] | Filter for data to be excluded from this processor's actions | no
exclude > regexp | [regexp][] | Regex cache settings. | no
exclude > attribute | [attribute][] | A list of attributes to match against. | no
exclude > resource | [resource][] | A list of items to match the resources against. | no
exclude > library | [library][] | A list of items to match the implementation library against. | no
exclude > log_severity_number | [log_severity_number][] | How to match against a log record's SeverityNumber, if defined. | no
```

> 

Reviewer: You're missing the sentence which explains the meaning of the `>` symbol: https://github.com/grafana/agent/blob/main/docs/developer/writing-flow-component-documentation.md#blocks

[output]: #output-block
[action]: #action-block
[include/exclude]: #include/exclude-blocks
[regexp]: #regexp-block
[attribute]: #attribute-block
[resource]: #resource-block
[library]: #library-block
[log_severity_number]: #log_severity_number-block

### action block

The `action` block configures how to modify the span, log or metric.

The following attributes are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`key` | `string` | The attribute that the action relates to. |  | yes
`action` | `string` | The type of action to be performed. |  | yes
`value` | `any` | The value to populate for the key. |  | no
`pattern` | `string` | A regex pattern. | `""` | no
`from_attribute` | `string` | The attribute from the input data to use to populate the attribute value. | `""` | no
`from_context` | `string` | The context value to use to populate the attribute value.  | `""` | no
`converted_type` | `string` | The type to convert the attribute value to. | `""` | no

> The type of `value` could be a number, a string or a boolean.

Reviewer: ```suggestion
The type of `value` must be either a number, string, or boolean.
```

The supported values for `action` are:

* `insert`: Inserts a new attribute in input data where the key does not already exist.

    * `key` is required. It specifies the attribute to act upon.
    * One of `value`, `from_attribute` or `from_context` is required.

>     * `action = "insert"` is required.

Reviewer: `action = "insert"` being required when `action` is `insert` is a little confusing to me. Can we drop this? 

Ditto with the other instances of this below (`action = "upsert"`, etc)

* `update`: Updates an attribute in input data where the key does exist.

    * `key` is required. It specifies the attribute to act upon.
    * One of `value`, `from_attribute` or `from_context` is required.
    * `action = "update"` is required.

> * `upsert`: Performs insert or update. Inserts a new attribute in input data where the key does not already exist and updates an attribute in input data where the key does exist.

Reviewer: ```suggestion
* `upsert`: Either inserts a new attribute in input data where the key does not already exist, or updates an attribute in input data where the key does exist.
```

* `hash`: Hashes (SHA1) an existing attribute value.

    * `key` and/or `pattern` is required.
    * `action = "hash"` is required.

* `extract`: Extracts values using a regular expression rule from the input key to target keys specified in the rule. If a target key already exists, it will be overridden. Note: It behaves similar to the Span Processor `to_attributes` setting with the existing attribute as the source.

    * `key` is required. It specifies the attribute to extract values from. The value of `key` is NOT altered.
    * `pattern` is required. It is the regex pattern used to extract attributes from the value of `key`. The submatchers must be named. If attributes already exist, they will be overwritten.
    * `action = "extract"` is required.

* `convert`: Converts an existing attribute to a specified type.

    * `key` is required. It specifies the attribute to act upon.
    * `action = "convert"` is required.
    * `converted_type` is required and must be one of int, double or string.

* `delete`: Deletes an attribute from the input data.

    * `key` and/or `pattern` is required. It specifies the attribute to act upon.
    * `action = "delete"` is required.

> ### include/exclude blocks

Author: Also, title "include/exclude blocks" could be misinterpreted as "how to include or exclude blocks".

> ### include/exclude blocks

Reviewer: > I don't want to have a ### include block and a ### exclude block separately, and copy paste the same contents.

I would prefer having both an `include` and `exclude` block; we should be optimizing for the reading experience over the writing experience. 

Instead of copying and pasting, you can put the content in docs/shared and use the docs/shared shortcode to deduplicate.

> ### include/exclude blocks

Author: It is not trivial to reuse code by putting it in `docs/shared`, because the "include block" and "exclude block" sections contain references to blocks such as `[attribute][]` and `[resource][]`. 

If we were to put the shared code in `docs/shared`, we could do one of those things:
1. Unlink the links in the shared code and leave them as text. This means changing `[attribute][]` to `attribute`. I think this wouldn't be a good solution, because the `attribute` block is described in the "Blocks" section. So the config in `docs/shared` for the "include block" would not be self sufficient.
2. Move the text with links to the "Blocks" section. This also wouldn't make sense, since the attributes in the "include" block are not mentioned yet. 
3. Move only the table to a shared md file, to avoid some repetition. But copy paste the bullet points about "One of the following is also required:"

Basically the problem is that in the "include block" description, we have to mention both its attributes and its blocks. But because the blocks definition is not in the "include block, it is hard to reuse md code.

For now I'll go with option 3, but please let me know if there is a better way.

The fact that blocks and attributes are defined separately in config suggests that they should be described separately, but in this case I cannot think of a reason why describing them in one place is wrong, and I can't think of a way to describe them separately.

The `include` and `exclude` blocks provide an option to include or exclude data from being fed into the [action] blocks, based on the properties of a span, log, or metric records.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`match_type` | `string` | Controls how items in "services" and "span_names" arrays are interpreted. | | yes
`services` | `list(string)` | A list of items to match the service name against. | `[]` | no
`span_names` | `list(string)` | A list of items to match the span name against. | `[]` | no
`log_bodies` | `list(string)` | A list of strings that the LogRecord's body field must match against. | `[]` | no
`log_severity_texts` | `list(string)` | A list of strings that the LogRecord's severity text field must match against. | `[]` | no
`metric_names` | `list(string)` | A list of strings to match the metric name against. | `[]` | no
`span_kinds` | `list(string)` | A list of items to match the span kind against. | `[]` | no

> `match_type` is required and can be set to either `"regexp"` or `"strict"`.

Reviewer: ```suggestion
`match_type` is required and must be set to either `"regexp"` or `"strict"`.
```

> One of the following is also required:

Author: The Otel docs also mentions [expressions](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.63.0/processor/attributesprocessor#includeexclude-filtering). However, I don't see any "expressions" in the code, so I removed the reference to it here.

Also, the Otel docs didn't mention "span_kinds", which I included here.

I could raise a PR to the Otel repo to also fix it on their end?

Reviewer: > I could raise a PR to the Otel repo to also fix it on their end?

Raising it upstream would be good OSS citizen behavior 👍

Author: Ok great, I raised a [PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20855) :)

> * For spans, one of `services`, `span_names`, `span_kinds`, [attribute], [resource], or [library] must be specified with a non-empty value for a valid configuration. The `log_bodies`, `log_severity_texts`, `resource_attributes` and `metric_names` fields are invalid.
> * For logs, one of `log_bodies`, `log_severity_texts`, [attribute], [resource], or [library] must be specified with a non-empty value for a valid configuration. The `span_names`, `span_kinds`, `metric_names`, `resource_attributes`, and `services` fields are invalid.
> * For metrics, one of `metric_names`, [resource] must be specified with a valid non-empty value for a valid configuration. The `span_names`, `span_kinds`, `log_bodies`, `log_severity_texts` and `services` fields are invalid.

Reviewer: ```suggestion
* For spans, one of `services`, `span_names`, `span_kinds`, [attribute][], [resource][], or [library][] must be specified with a non-empty value for a valid configuration. The `log_bodies`, `log_severity_texts`, `resource_attributes` and `metric_names` attributes are invalid.
* For logs, one of `log_bodies`, `log_severity_texts`, [attribute][], [resource][], or [library][] must be specified with a non-empty value for a valid configuration. The `span_names`, `span_kinds`, `metric_names`, `resource_attributes`, and `services` attributes are invalid.
* For metrics, one of `metric_names`, [resource][] must be specified with a valid non-empty value for a valid configuration. The `span_names`, `span_kinds`, `log_bodies`, `log_severity_texts` and `services` attributes are invalid.
```

### regexp block

This block is optional configuration for the `match_type` of `"regexp"`.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`cacheenabled` | `bool` | Determines whether match results are LRU cached. | `false` | no
`cachemaxnumentries` | `int` | The max number of entries of the LRU cache that stores match results. | `0` | no

Enabling `cacheenabled` could make subsequent matches faster.
Cache size is unlimited unless `cachemaxnumentries` is also specified.

`cachemaxnumentries` is ignored if `cacheenabled` is false.

### attribute block

This block specifies a list of attributes to match against.
Only `match_type = "strict"` is allowed if `attribute` is specified.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`key` | `string` | The attribute key. | | yes
`value` | `any` | The value to match against. | | no

If `value` is not set, any value will match.
The type of `value` could be a number, a string or a boolean.

> ### resource block

Author: This is the same as the attribute block. Should I repeat it, or try have it under the same heading ("attributes/resource block")?

Reviewer: No, don't merge blocks please, even if they have the same arguments.

This block specifies items to match the resources against.
A match occurs if the input data resources matches at least one `resource` block.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`key` | `string` | The attribute key. | | yes
`value` | `any` | The value to match against. | | no

If `value` is not set, any value will match.
The type of `value` could be a number, a string or a boolean.

### library block

This block specifies items to match the implementation library against.
A match occurs if the span's implementation library matches at least one `library` block.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`name` | `string` | The attribute key. | | yes

> `version` | | The value to match against. | | yes

Author: `version` is a pointer. Should we include a note in the docs that leaving this unset and setting it to an empty string are not the same thing, and could lead to different outcomes?

Reviewer: I don't think it has to be a pointer on our side, right? We can turn it into a pointer when converting to the upstream type.

Author: The matching rules for null pointers and empty strings work differently:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/filter/filterconfig/config.go#L214-L226

If `version` is a null pointer, it will match any version. If `version` is an empty string, it will only match a version which is also an empty string.

I will try to add some unit tests to make sure this works. If it doesn't work, then I could probably add an extra bool flag in the river config so that our code sets the version to a null pointer during the conversion.

### output block

{{< docs/shared lookup="flow/reference/components/output-block.md" source="agent" >}}

## Exported fields

The following fields are exported and can be referenced by other components:

Name | Type | Description
---- | ---- | -----------
`input` | `otelcol.Consumer` | A value that other components can use to send telemetry data to.

`input` accepts `otelcol.Consumer` data for any telemetry signal (metrics, logs, or traces).

## Component health

`otelcol.processor.attributes` is only reported as unhealthy if given an invalid
configuration.

## Debug information

`otelcol.processor.attributes` does not expose any component-specific debug
information.

## Examples

### Various uses of the "action" block

```river
otelcol.receiver.otlp "default" {
  http {}
  grpc {}

  output {
    metrics = [otelcol.processor.attributes.default.input]
    logs    = [otelcol.processor.attributes.default.input]
    traces  = [otelcol.processor.attributes.default.input]
  }
}

> otelcol.processor.attributes "default" {

Author: Add an example for `from_context`?

        match_type = "regexp"
        // The span name must not match "login.*" pattern.
        span_names = ["login.*"]
    }

    action {
        key = "password"
        action = "update"
        value = "obfuscated"
    }

    action {
        key = "token"
        action = "delete"
    }

    output {
        metrics = [otelcol.exporter.otlp.default.input]
        logs    = [otelcol.exporter.otlp.default.input]
        traces  = [otelcol.exporter.otlp.default.input]
    }
}
```

### Including spans based on regex and attributes

The following demonstrates how to process spans that have an attribute that matches a regexp patterns.
This processor will obfuscate the "db.statement" attribute in spans where the "db.statement" attribute
matches a regex pattern.

```river
otelcol.processor.attributes "default" {
    include {
        // "match_type" of "regexp" defines that the "value" attributes 
        // in the "attribute" blocks are regexp-es.
        match_type = "regexp"

        // This attribute ('db.statement') must exist in the span and match 
        // the regex ('SELECT \* FROM USERS.*') for a match.
        attribute {

>             key = "env"

Author: Shouldn't key = "env" be key = "db.statement"? This was taken from the [examples](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.63.0/processor/attributesprocessor/testdata/config.yaml#L303) in the otel repo.

### resource block

This block specifies items to match the resources against.
A match occurs if the input data resources matches at least one `resource` block.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`key` | `string` | The attribute key. | | yes
`value` | `any` | The value to match against. | | no

If `value` is not set, any value will match.
The type of `value` could be a number, a string or a boolean.

### library block

This block specifies items to match the implementation library against.
A match occurs if the span's implementation library matches at least one `library` block.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`name` | `string` | The attribute key. | | yes
`version` | | The value to match against. | | yes

### log_severity_number block

This block defines how to match based on a log record's SeverityNumber field.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------

> `min` | `int` | The lowest severity that may be matched. | | yes

Author: Not sure what is the best way to document the allowed [plog values](https://github.com/open-telemetry/opentelemetry-collector/blob/v0.64.0/pdata/plog/logs.go#L76). I think I'll just copy all the values here, so that we don't link to the otel code.

Reviewer: Yes, you should copy the values. But also, I think we should define our own string-based mapping here using encoding.TextMarshaler and encoding.TextUnmarshaler because `min = 7` is going to confuse people.

* For logs, one of `log_bodies`, `log_severity_texts`, `log_severity`, [attribute][], [resource][] or [library][] must be specified with a non-empty value for a valid configuration. The `span_names`, `span_kinds`, `metric_names`, and `services` attributes are invalid.
* For metrics, one of `metric_names` or [resource][] must be specified with a valid non-empty value for a valid configuration. The `span_names`, `span_kinds`, `log_bodies`, `log_severity_texts`, `log_severity`, `services`, [attribute][] and [library][] attributes are invalid.

For `metric_names`, a match occurs if the metric name matches at least one item in the list.
For `span_kinds`, a match occurs if the span's span kind matches at least one item in the list.

### regexp block

This block is optional configuration for the `match_type` of `"regexp"`.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------

> `cacheenabled` | `bool` | Determines whether match results are LRU cached. | `false` | no

Reviewer 2: `cache_enabled` seems more inline with our naming. Same for `cache_max_num_entries`

Reviewer 2: Link to what LRU is  or spell it out?

* More than one `library` block can be defined.
* A match occurs if the span's implementation library matches at least one `library` block.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`name` | `string` | The attribute key. | | yes
`version` | `string` | The version to match against. | null | no

If `version` is unset, any version will match. If `version` is set to an empty string, it will only match 
a library version which is also an empty string.

### log_severity block

This block defines how to match based on a log record's SeverityNumber field.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`min` | `string` | The lowest severity that may be matched. | | yes
`match_undefined` | `bool` | Whether logs with "undefined" severity match. | | yes

If `match_undefined` is true, entries with undefined severity will match.

The severities supported by Otel are listed in the table below. 
The value for `min` should be one of the values in the "Log Severity" column.

> Log Severity | Severity number

Reviewer 2: This feels a bit strange but assuming its a straight mapping from otel right?
